File: 23-merge-k-sorted-lists/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* curr = new ListNode(0);
        ListNode* list3 = curr;
        while(list1 != NULL || list2 != NULL) {
            int x = list1 != NULL ? list1->val : INT_MAX;
            int y = list2 != NULL ? list2->val : INT_MAX;

            if(x < y) {
                list3->next = list1;
                list1 = list1 != NULL ? list1->next : NULL;
                
            }else{
                list3->next = list2;
                list2 = list2 != NULL ? list2->next : NULL;
            }
            
            list3 = list3->next;

        }
        return curr->next;
        
    }

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size() == 0) return nullptr;
        if(lists.size() == 1) return lists[0];

        ListNode* l1 = lists[0];
        for(int i=1;i<lists.size();i++){
            ListNode* l2 = lists[i];
            l1 = mergeTwoLists(l1,l2);
        }

        return l1;
        
    }
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> q = new PriorityQueue<>(
            (l1,l2) -> Integer.compare(l1.val , l2.val));

            for(ListNode l1: lists){
                if(l1 != null){
                    q.offer(l1);
                }
            }

            ListNode dummy = new ListNode();
            ListNode l3 = dummy;

            while(!q.isEmpty()){
                ListNode curr = q.poll();
                l3.next = curr;
                l3 = l3.next;
                if(curr.next != null){
                    q.offer(curr.next);
                }
            }

        
        return dummy.next;
    }
}